README.md
# `pbrocks-env`

`pbrocks-env` lets you easily set up a local WordPress environment for building and testing plugins and themes. It's simple to install and requires no configuration.

## Quick (tl;dr) instructions

Ensure that Docker is running, then:

```sh
$ cd /path/to/a/wordpress/plugin
$ npm -g i @wordpress/env
$ pbrocks-env start
```

The local environment will be available at http://localhost:8888 (Username: `admin`, Password: `password`).

The database credentials are: user `root`, password `password`. For a comprehensive guide on connecting directly to the database, refer to [Accessing the MySQL Database](https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/getting-started-with-code-contribution.md#accessing-the-mysql-database).

## Prerequisites

`pbrocks-env` requires Docker to be installed. There are instructions available for installing Docker on [Windows](https://docs.docker.com/desktop/install/windows-install/), [macOS](https://docs.docker.com/docker-for-mac/install/), and [Linux](https://docs.docker.com/desktop/install/linux-install/).

Node.js and npm are required. The latest LTS version of Node.js is used to develop `pbrocks-env` and is recommended.

## Installation

### Installation as a global package

After confirming that the prerequisites are installed, you can install `pbrocks-env` globally like so:

```sh
$ npm -g i @wordpress/env
```

You're now ready to use `pbrocks-env`!

### Installation as a local package

If your project already has a package.json, it's also possible to use `pbrocks-env` as a local package. First install `pbrocks-env` locally as a dev dependency:

```sh
$ npm i @wordpress/env --save-dev
```

At this point, you can use the local, project-level version of pbrocks-env via [`npx`](https://www.npmjs.com/package/npx), a utility automatically installed with `npm`.`npx` finds binaries like pbrocks-env installed through node modules. As an example: `npx pbrocks-env start --update`.

If you don't wish to use `npx`, modify your package.json and add an extra command to npm `scripts` (https://docs.npmjs.com/misc/scripts):

```json
"scripts": {
	"pbrocks-env": "pbrocks-env"
}
```

When installing `pbrocks-env` in this way, all `pbrocks-env` commands detailed in these docs must be prefixed with `npm run`, for example:

```sh
# You must add another double dash to pass flags to the script (pbrocks-env) rather than to npm itself
$ npm run pbrocks-env start -- --update
```

instead of:

```sh
$ pbrocks-env start --update
```

## Usage

### Starting the environment

First, ensure that Docker is running. You can do this by clicking on the Docker icon in the system tray or menu bar.

Then, change to a directory that contains a WordPress plugin or theme:

```sh
$ cd ~/gutenberg
```

Then, start the local environment:

```sh
$ pbrocks-env start
```

Finally, navigate to http://localhost:8888 in your web browser to see WordPress running with the local WordPress plugin or theme running and activated. Default login credentials are username: `admin` password: `password`.

### Stopping the environment

To stop the local environment:

```sh
$ pbrocks-env stop
```

## Troubleshooting common problems

Many common problems can be fixed by running through the following troubleshooting steps in order:

### 1. Check that `pbrocks-env` is running

First, check that `pbrocks-env` is running. One way to do this is to have Docker print a table with the currently running containers:

```sh
$ docker ps
```

In this table, by default, you should see three entries: `wordpress` with port 8888, `tests-wordpress` with port 8889 and `mariadb` with port 3306.

### 2. Check the port number

By default `pbrocks-env` uses port 8888, meaning that the local environment will be available at http://localhost:8888.

You can configure the port that `pbrocks-env` uses so that it doesn't clash with another server by specifying the `WP_ENV_PORT` environment variable when starting `pbrocks-env`:

```sh
$ WP_ENV_PORT=3333 pbrocks-env start
```

Running `docker ps` and inspecting the `PORTS` column allows you to determine which port `pbrocks-env` is currently using.

You may also specify the port numbers in your `.pbrocks-env.json` file, but the environment variables take precedent.

### 3. Restart `pbrocks-env` with updates

Restarting `pbrocks-env` will restart the underlying Docker containers which can fix many issues.

To restart `pbrocks-env`, just run `pbrocks-env start` again. It will automatically stop and start the container. If you also pass the `--update` argument, it will download updates and configure WordPress again.

```sh
$ pbrocks-env start --update
```

### 4. Restart Docker

Restarting Docker will restart the underlying Docker containers and volumes which can fix many issues.

To restart Docker:

1. Click on the Docker icon in the system tray or menu bar.
2. Select _Restart_.

Once restarted, start `pbrocks-env` again:

```sh
$ pbrocks-env start
```

### 5. Reset the database

Resetting the database which the local environment uses can fix many issues, especially when they are related to the WordPress installation.

To reset the database:

**⚠️ WARNING: This will permanently delete any posts, pages, media, etc. in the local WordPress installation.**

```sh
$ pbrocks-env clean all
$ pbrocks-env start
```

### 6. Destroy everything and start again 🔥

When all else fails, you can use `pbrocks-env destroy` to forcibly remove all of the underlying Docker containers, volumes, and files. This will allow you to start from scratch.

To do so:

**⚠️ WARNING: This will permanently delete any posts, pages, media, etc. in the local WordPress installation.**

```sh
$ pbrocks-env destroy
# This new instance is a fresh start with no existing data:
$ pbrocks-env start
```

### 7. Debug mode and inspecting the generated dockerfile.

`pbrocks-env` uses docker behind the scenes. Inspecting the generated docker-compose file can help to understand what's going on.

Start `pbrocks-env` in debug mode

```sh
pbrocks-env start --debug
```

`pbrocks-env` will output its config which includes `dockerComposeConfigPath`.

```sh
ℹ Config:
	...
	"dockerComposeConfigPath": "/Users/$USERNAME/.pbrocks-env/5a619d332a92377cd89feb339c67b833/docker-compose.yml",
	...
```

## Using included WordPress PHPUnit test files

Out of the box `pbrocks-env` includes the [WordPress' PHPUnit test files](https://develop.svn.wordpress.org/trunk/tests/phpunit/) corresponding to the version of WordPress installed. There is an environment variable, `WP_TESTS_DIR`, which points to the location of these files within each container. By including these files in the environment, we remove the need for you to use a package or install and mount them yourself. If you do not want to use these files, you should ignore the `WP_TESTS_DIR` environment variable and load them from the location of your choosing.

### Customizing the `wp-tests-config.php` file

While we do provide a default `wp-tests-config.php` file within the environment, there may be cases where you want to use your own. WordPress provides a `WP_TESTS_CONFIG_FILE_PATH` constant that you can use to change the `wp-config.php` file used for testing. Set this to a desired path in your `bootstrap.php` file and the file you've chosen will be used instead of the one included in the environment.

## Using Xdebug

Xdebug is installed in the pbrocks-env environment, but it is turned off by default. To enable Xdebug, you can use the `--xdebug` flag with the `pbrocks-env start` command. Here is a reference to how the flag works:

```sh
# Sets the Xdebug mode to "debug" (for step debugging):
pbrocks-env start --xdebug

# Sets the Xdebug mode to "off":
pbrocks-env start

# Enables each of the Xdebug modes listed:
pbrocks-env start --xdebug=profile,trace,debug
```

When you're running `pbrocks-env` using `npm run`, like when working in the Gutenberg repo or when `pbrocks-env` is a local project dependency, don't forget to add an extra double dash before the `--xdebug` command:

```sh
npm run pbrocks-env start -- --xdebug
# Alternatively, use npx:
npx pbrocks-env start --xdebug
```

If you forget about that, the `--xdebug` parameter will be passed to npm instead of the `pbrocks-env start` command and it will be ignored.

You can see a reference on each of the Xdebug modes and what they do in the [Xdebug documentation](https://xdebug.org/docs/all_settings#mode).

_Since we are only installing Xdebug 3, Xdebug is only supported for PHP versions greater than or equal to 7.2 (the default). Xdebug won't be installed if `phpVersion` is set to a legacy version._

### Xdebug IDE support

To connect to Xdebug from your IDE, you can use these IDE settings. This bit of JSON was tested for VS Code's `launch.json` format (which you can [learn more about here](https://code.visualstudio.com/docs/editor/debugging#_launchjson-attributes)) along with [this PHP Debug extension](https://marketplace.visualstudio.com/items?itemName=felixfbecker.php-debug). Its path mapping also points to a specific plugin -- you should update this to point to the source you are working with inside of the pbrocks-env instance.

You should only have to translate `port` and `pathMappings` to the format used by your own IDE.

```json
{
  "name": "Listen for XDebug",
  "type": "php",
  "request": "launch",
  "port": 9003,
  "pathMappings": {
    "/var/www/html/wp-content/plugins/gutenberg": "${workspaceFolder}/"
  }
}
```

After you create a `.vscode/launch.json` file in your repository, you probably want to add it to your [global gitignore file](https://docs.github.com/en/get-started/getting-started-with-git/ignoring-files#configuring-ignored-files-for-all-repositories-on-your-computer) so that it stays private for you and is not committed to the repository.

Once your IDEs Xdebug settings have been enabled, you should just have to launch the debugger, put a breakpoint on any line of PHP code, and then refresh your browser!

Here is a summary:

1. Start pbrocks-env with xdebug enabled: `pbrocks-env start --xdebug`
2. Install a suitable Xdebug extension for your IDE if it does not include one already.
3. Configure the IDE debugger to use port `9003` and the correct source files in pbrocks-env.
4. Launch the debugger and put a breakpoint on any line of PHP code.
5. Refresh the URL pbrocks-env is running at and the breakpoint should trigger.

## Command reference

`pbrocks-env` creates generated files in the `pbrocks-env` home directory. By default, this is `~/.pbrocks-env`. The exception is Linux, where files are placed at `~/pbrocks-env` [for compatibility with Snap Packages](https://github.com/WordPress/gutenberg/issues/20180#issuecomment-587046325). The `pbrocks-env` home directory contains a subdirectory for each project named `/$md5_of_project_path`. To change the `pbrocks-env` home directory, set the `WP_ENV_HOME` environment variable. For example, running `WP_ENV_HOME="something" pbrocks-env start` will download the project files to the directory `./something/$md5_of_project_path` (relative to the current directory).

### `pbrocks-env start`

The start command installs and initializes the WordPress environment, which includes downloading any specified remote sources. By default, `pbrocks-env` will not update or re-configure the environment except when the configuration file changes. Tell `pbrocks-env` to update sources and apply the configuration options again with `pbrocks-env start --update`. This will not overwrite any existing content.

```sh
pbrocks-env start

Starts WordPress for development on port 8888 (override with WP_ENV_PORT) and
tests on port 8889 (override with WP_ENV_TESTS_PORT). The current working
directory must be a WordPress installation, a plugin, a theme, or contain a
.pbrocks-env.json file. After first install, use the '--update' flag to download
updates to mapped sources and to re-apply WordPress configuration options.

Options:
  --help     Show help                                                 [boolean]
  --version  Show version number                                       [boolean]
  --debug    Enable debug output.                     [boolean] [default: false]
  --update   Download source updates and apply WordPress configuration.
                                                      [boolean] [default: false]
  --xdebug   Enables Xdebug. If not passed, Xdebug is turned off. If no modes
             are set, uses "debug". You may set multiple Xdebug modes by passing
             them in a comma-separated list: `--xdebug=develop,coverage`. See
             https://xdebug.org/docs/all_settings#mode for information about
             Xdebug modes.                                              [string]
```

### `pbrocks-env stop`

```sh
pbrocks-env stop

Stops running WordPress for development and tests and frees the ports.
```

### `pbrocks-env clean [environment]`

```sh
pbrocks-env clean [environment]

Cleans the WordPress databases.

Positionals:
  environment  Which environments' databases to clean.
            [string] [choices: "all", "development", "tests"] [default: "tests"]
```

### `pbrocks-env run [container] [command]`

The run command can be used to open shell sessions or invoke WP-CLI commands.

<div class="callout callout-alert">
In some cases, `pbrocks-env` may consume options that you are attempting to pass to
the container. This happens with options that `pbrocks-env` has already declared,
such as `--debug`, `--help`, and `--version`. When this happens, you should fall
back to using quotation marks; `pbrocks-env` considers everything inside the
quotation marks to be command argument.

For example, to ask `WP-CLI` for its help text:

```sh
pbrocks-env run cli "wp --help"
```

Without the quotation marks, `pbrocks-env` will print its own help text instead of
passing it to the container. If you experience any problems where the command
is not being passed correctly, fall back to using quotation marks.
</div>

```sh
pbrocks-env run <container> [command..]

Runs an arbitrary command in one of the underlying Docker containers. The
"container" param should reference one of the underlying Docker services like
"development", "tests", or "cli". To run a wp-cli command, use the "cli" or
"tests-cli" service. You can also use this command to open shell sessions like
bash and the WordPress shell in the WordPress instance. For example, `pbrocks-env run
cli bash` will open bash in the development WordPress instance.

Positionals:
  container  The container to run the command on.            [string] [required]
  command    The command to run.                           [array] [default: []]

Options:
  --help     Show help                                                 [boolean]
  --version  Show version number                                       [boolean]
  --debug    Enable debug output.                     [boolean] [default: false]
```

For example:

#### Displaying the users on the development instance:

```sh
pbrocks-env run cli wp user list
⠏ Running `wp user list` in 'cli'.

ID      user_login      display_name    user_email      user_registered roles
1       admin   admin   wordpress@example.com   2020-03-05 10:45:14     administrator

✔ Ran `wp user list` in 'cli'. (in 2s 374ms)
```

#### Creating a post on the tests instance:

```sh
pbrocks-env run tests-cli "wp post create --post_type=page --post_title='Ready'"

ℹ Starting 'wp post create --post_type=page --post_title='Ready'' on the tests-cli container.

Success: Created post 5.
✔ Ran `wp post create --post_type=page --post_title='Ready'` in 'tests-cli'. (in 3s 293ms)
```

#### Opening the WordPress shell on the tests instance and running PHP commands:

```sh
pbrocks-env run tests-cli wp shell
ℹ Starting 'wp shell' on the tests-cli container. Exit the WordPress shell with ctrl-c.

Starting 31911d623e75f345e9ed328b9f48cff6_mysql_1 ... done
Starting 31911d623e75f345e9ed328b9f48cff6_tests-wordpress_1 ... done
wp> echo( 'hello world!' );
hello world!
wp> ^C
✔ Ran `wp shell` in 'tests-cli'. (in 16s 400ms)
```

#### Installing a plugin or theme on the development instance

```sh
pbrocks-env run cli wp plugin install custom-post-type-ui

Creating 500cd328b649d63e882d5c4695871d04_cli_run ... done
Installing Custom Post Type UI (1.9.2)
Downloading installation package from https://downloads.wordpress.org/plugin/custom-post-type-ui.zip...
The authenticity of custom-post-type-ui.zip could not be verified as no signature was found.
Unpacking the package...
Installing the plugin...
Plugin installed successfully.
Success: Installed 1 of 1 plugins.
✔ Ran `plugin install custom-post-type-ui` in 'cli'. (in 6s 483ms)
```

**NOTE**: Depending on your host OS, you may experience errors when trying to install plugins or themes (e.g. `Warning: Could not create directory.`). This is typically because the user ID used within the container does not have write access to the mounted directories created by `pbrocks-env`. To resolve this, run the `docker-compose` command directly from the directory created by `pbrocks-env` and add `-u $(id -u)` and `-e HOME=/tmp` the `run` command as options:

```sh
$ cd ~/pbrocks-env/500cd328b649d63e882d5c4695871d04
$ docker-compose run --rm -u $(id -u) -e HOME=/tmp cli [plugin|theme] install <plugin|theme>
```

### `pbrocks-env destroy`

```sh
pbrocks-env destroy

Destroy the WordPress environment. Deletes docker containers, volumes, and
networks associated with the WordPress environment and removes local files.
```

### `pbrocks-env logs [environment]`

```sh
pbrocks-env logs

displays PHP and Docker logs for given WordPress environment.

Positionals:
  environment  Which environment to display the logs from.
      [string] [choices: "development", "tests", "all"] [default: "development"]

Options:
  --help     Show help                                                 [boolean]
  --version  Show version number                                       [boolean]
  --debug    Enable debug output.                     [boolean] [default: false]

CHANGELOG.md
<!-- Learn how to maintain this file at https://github.com/WordPress/gutenberg/tree/HEAD/packages#maintaining-changelogs. -->

## Unreleased

## 5.12.0 (2023-02-15)

## 5.11.0 (2023-02-01)

### Bug fix

-   PHP 7.3 and 7.4 must use PHPUnit 9.

### Enhancement

-   It's now possible to run PHPUnit tests on PHP 8.1 and 8.2.

## 5.10.0 (2023-01-11)

## 5.9.0 (2023-01-02)

## 5.8.0 (2022-12-14)

## 5.7.0 (2022-11-16)

## 5.6.0 (2022-11-02)

## 5.5.0 (2022-10-19)

## 5.4.0 (2022-10-05)

## 5.3.0 (2022-09-21)

## 5.2.0 (2022-08-16)

### Enhancement
-   Query parameters can now be used in .zip source URLs.

## 5.1.1 (2022-08-16)

### Bug Fix
-   Fix a crash when "core" was set to `null` in a `.pbrocks-env.json` file. We now use the latest stable WordPress version in that case. This also restores the previous behavior of `"core": null` in `.pbrocks-env.override.json`, which was to use the latest stable WordPress version.

## 5.1.0 (2022-08-10)

### Enhancement
-   Previously, pbrocks-env used the WordPress version provided by Docker in the WordPress image for installations which don't specify a WordPress version. Now, pbrocks-env will find the latest stable version on WordPress.org and check out the https://github.com/WordPress/WordPress repository at the tag matching that version. In most cases, this will match what Docker provides. The benefit is that pbrocks-env (and WordPress.org) now controls the default WordPress version rather than Docker.

### Bug Fix
-   Downloading a default WordPress version also resolves a bug where the wrong WordPress test files were used if no core source was specified in pbrocks-env.json. The current trunk test files were downloaded rather than the stable version. Now, the test files will match the default stable version.

## 5.0.0 (2022-07-27)

### Breaking Changes
-   Removed the `WP_PHPUNIT__TESTS_CONFIG` environment variable from the `phpunit` container. **This removes automatic support for the `wp-phpunit/wp-phpunit` Composer package. To continue using the package, set the following two environment variables in your `phpunit.xml` file or similar: `WP_TESTS_DIR=""` and `WP_PHPUNIT__TESTS_CONFIG="/wordpress-phpunit/wp-tests-config.php"`.**
-   Removed the generated `/var/www/html/phpunit-wp-config.php` file from the environment.

### Enhancement
-   Read WordPress' version and include the corresponding PHPUnit test files in the environment.
-   Set the `WP_TESTS_DIR` environment variable in all containers to point at the PHPUnit test files.

### Bug Fix
-   Restrict `WP_TESTS_DOMAIN` constant to just hostname rather than an entire URL (e.g. it now excludes scheme, port, etc.) ([#41039](https://github.com/WordPress/gutenberg/pull/41039)).

## 4.8.0 (2022-06-01)

### Enhancement
-   Removed the need for quotation marks when passing options to `pbrocks-env run`.
-   Setting a `config` key to `null` will prevent adding the constant to `wp-config.php` even if a default value is defined by `pbrocks-env`.

## 4.7.0 (2022-05-18)

### Enhancement
-   Added SSH protocol support for git sources

## 4.2.0 (2022-01-27)

### Enhancement
-   Added command `pbrocks-env install-path` to list the directory used for the environment.
-   The help entry is now shown when no subcommand is passed to `pbrocks-env`.

### Bug Fix
-   Updated `yargs` to fix [CVE-2021-3807](https://nvd.nist.gov/vuln/detail/CVE-2021-3807).

## 4.1.3 (2021-11-07)

### Bug Fix

-   Fix Xdebug installation code to ensure it would fail gracefully

## 4.0.3 (2021-04-29)

### Bug Fix

-   `pbrocks-env destroy` will now work in environments which don't include the `grep` or `awk` commands, such as Windows PowerShell.
-   Fix several permissions issues related to wp-config.php and wp-content files.
-   Fix crash which happened when the path to pbrocks-env's home directory contained a space.
-   Disable Xdebug 3 for PHP versions less than 7.2 to resolve startup crash.

## 4.0.0 (2021-03-17)

### Breaking Change

-   Migrate from `nodegit` to `simple-git`. This change now requires you to have a `git` binary installed locally to utilize the git sources feature of pbrocks-env.json.

### Bug Fixes

-   "mappings" sources are now downloaded if they contain non-local sources.

## 3.0.0 (2020-12-17)

### Breaking Changes

-   `pbrocks-env start` is now the only command which writes to the docker configuration files. Previously, running any command would also parse the config and then write it to the correct location. Now, other commands still parse the config, but they will not overwrite the confugiration which was set by pbrocks-env start. This allows parameters to be passed to pbrocks-env start which can affect the configuration.

### Enhancement

-   Update nodegit dependency to 0.27.0, the earlier version does not have pre-built binaries for Node 14.15.0 LTS. Upgrading provides support without requiring building nodegit locally.
-   Allow WP_HOME wp-config value to be set to a custom port other than the default for the docker instance.
-   Append the instance URL to output of `pbrocks-env start`.

### New feature

-   Add support for setting the PHP version used for the WordPress instance. For example, test PHP 8 with `"phpVersion": 8.0` in pbrocks-env.json.
-   Add Xdebug 3 to the development environment. You can enable Xdebug with `pbrocks-env start --xdebug` (for debug mode) or `pbrocks-env start --xdebug=develop,coverage` for custom modes.

### Bug Fixes

-   ZIP-based plugin sources are now downloaded to a directory using the basename of the URL instead of the full URL path. This prevents HTML encoded characters in the URL (like "/") from being improperly encoded into the filesystem. This fixes the issue where many .zip sources broke because files with these badly formatted characters were not loaded as assets.

## 2.0.0 (2020-09-03)

### Breaking Changes

-   The `config` and `mappings` options in `.pbrocks-env.json` are now merged with any overrides instead of being overwritten.

lib/config/config.js
'use strict';
/**
 * External dependencies
 */
const fs = require( 'fs' ).promises;
const path = require( 'path' );
const os = require( 'os' );

/**
 * Internal dependencies
 */
const detectDirectoryType = require( './detect-directory-type' );
const { validateConfig, ValidationError } = require( './validate-config' );
const readRawConfigFile = require( './read-raw-config-file' );
const parseConfig = require( './parse-config' );
const { includeTestsPath, parseSourceString } = parseConfig;
const md5 = require( '../md5' );

/**
 * pbrocks-env configuration.
 *
 * @typedef WPConfig
 * @property {string}                           name                    Name of the environment.
 * @property {string}                           configDirectoryPath     Path to the .pbrocks-env.json file.
 * @property {string}                           workDirectoryPath       Path to the work directory located in ~/.pbrocks-env.
 * @property {string}                           dockerComposeConfigPath Path to the docker-compose.yml file.
 * @property {boolean}                          detectedLocalConfig     If true, pbrocks-env detected local config and used it.
 * @property {Object.<string, WPServiceConfig>} env                     Specific config for different environments.
 * @property {boolean}                          debug                   True if debug mode is enabled.
 */

/**
 * Base-level config for any particular environment. (development/tests/etc)
 *
 * @typedef WPServiceConfig
 * @property {WPSource}                  coreSource    The WordPress installation to load in the environment.
 * @property {WPSource[]}                pluginSources Plugins to load in the environment.
 * @property {WPSource[]}                themeSources  Themes to load in the environment.
 * @property {number}                    port          The port to use.
 * @property {Object}                    config        Mapping of wp-config.php constants to their desired values.
 * @property {Object.<string, WPSource>} mappings      Mapping of WordPress directories to local directories which should be mounted.
 * @property {string}                    phpVersion    Version of PHP to use in the environments, of the format 0.0.
 */

/**
 * A WordPress installation, plugin or theme to be loaded into the environment.
 *
 * @typedef WPSource
 * @property {'local'|'git'|'zip'} type     The source type.
 * @property {string}              path     The path to the WordPress installation, plugin or theme.
 * @property {?string}             url      The URL to the source download if the source type is not local.
 * @property {?string}             ref      The git ref for the source if the source type is 'git'.
 * @property {string}              basename Name that identifies the WordPress installation, plugin or theme.
 */

/**
 * Reads, parses, and validates the given .pbrocks-env.json file into a pbrocks-env config
 * object for internal use.
 *
 * @param {string} configPath Path to the .pbrocks-env.json file.
 *
 * @return {WPConfig} A parsed and validated pbrocks-env config object.
 */
module.exports = async function readConfig( configPath ) {
	const configDirectoryPath = path.dirname( configPath );
	const workDirectoryPath = path.resolve(
		await getHomeDirectory(),
		md5( configPath )
	);

	// The specified base configuration from .pbrocks-env.json or from the local
	// source type which was automatically detected.
	const baseConfig =
		( await readRawConfigFile( '.pbrocks-env.json', configPath ) ) ||
		( await getDefaultBaseConfig( configPath ) );

	// Overriden .pbrocks-env.json on a per-user case.
	const overrideConfig =
		( await readRawConfigFile(
			'.pbrocks-env.override.json',
			configPath.replace( /\.pbrocks-env\.json$/, '.pbrocks-env.override.json' )
		) ) || {};

	const detectedLocalConfig =
		Object.keys( { ...baseConfig, ...overrideConfig } ).length > 0;

	// Default configuration which is overridden by .pbrocks-env.json files.
	const defaultConfiguration = {
		core: null, // Indicates that the latest stable version should ultimately be used.
		phpVersion: null,
		plugins: [],
		themes: [],
		port: 8888,
		mappings: {},
		config: {
			WP_DEBUG: true,
			SCRIPT_DEBUG: true,
			WP_ENVIRONMENT_TYPE: 'local',
			WP_PHP_BINARY: 'php',
			WP_TESTS_EMAIL: 'admin@example.org',
			WP_TESTS_TITLE: 'Test Blog',
			WP_TESTS_DOMAIN: 'localhost',
			WP_SITEURL: 'http://localhost',
			WP_HOME: 'http://localhost',
		},
		env: {
			development: {}, // No overrides needed, but it should exist.
			tests: {
				config: { WP_DEBUG: false, SCRIPT_DEBUG: false },
				port: 8889,
			},
		},
	};

	// A quick validation before merging on a service by service level allows us
	// to check the root configuration options and provide more helpful errors.
	validateConfig(
		mergeWpServiceConfigs( [
			defaultConfiguration,
			baseConfig,
			overrideConfig,
		] )
	);

	// A unique array of the environments specified in the config options.
	// Needed so that we can override settings per-environment, rather than
	// overwriting each environment key.
	const getEnvKeys = ( config ) => Object.keys( config.env || {} );
	const allEnvs = [
		...new Set( [
			...getEnvKeys( defaultConfiguration ),
			...getEnvKeys( baseConfig ),
			...getEnvKeys( overrideConfig ),
		] ),
	];

	// Returns a pair with the root config options and the specific environment config options.
	const getEnvConfig = ( config, envName ) => [
		config,
		config.env && config.env[ envName ] ? config.env[ envName ] : {},
	];

	// Merge each of the specified environment-level overrides.
	const allPorts = new Set(); // Keep track of unique ports for validation.
	const env = {};
	for ( const envName of allEnvs ) {
		env[ envName ] = await parseConfig(
			validateConfig(
				mergeWpServiceConfigs( [
					...getEnvConfig( defaultConfiguration, envName ),
					...getEnvConfig( baseConfig, envName ),
					...getEnvConfig( overrideConfig, envName ),
				] ),
				envName
			),
			{ workDirectoryPath }
		);
		allPorts.add( env[ envName ].port );
	}

	if ( allPorts.size !== allEnvs.length ) {
		throw new ValidationError(
			'Invalid .pbrocks-env.json: Each port value must be unique.'
		);
	}

	return withOverrides( {
		name: path.basename( configDirectoryPath ),
		dockerComposeConfigPath: path.resolve(
			workDirectoryPath,
			'docker-compose.yml'
		),
		configDirectoryPath,
		workDirectoryPath,
		detectedLocalConfig,
		env,
	} );
};

/**
 * Deep-merges the values in the given service environment. This allows us to
 * merge the wp-config.php values instead of overwriting them. Note that this
 * merges configs before they have been validated, so the passed config shape
 * will not match the WPServiceConfig type.
 *
 * @param {Object[]} configs Array of raw service config objects to merge.
 *
 * @return {Object} The merged configuration object.
 */
function mergeWpServiceConfigs( configs ) {
	// Returns an array of nested values in the config object. For example,
	// an array of all the wp-config objects.
	const mergeNestedObjs = ( key ) =>
		Object.assign(
			{},
			...configs.map( ( config ) => {
				if ( ! config[ key ] ) {
					return {};
				} else if ( typeof config[ key ] === 'object' ) {
					return config[ key ];
				}
				throw new ValidationError(
					`Invalid .pbrocks-env.json: "${ key }" must be an object.`
				);
			} )
		);

	const mergedConfig = {
		...Object.assign( {}, ...configs ),
		config: mergeNestedObjs( 'config' ),
		mappings: mergeNestedObjs( 'mappings' ),
	};

	delete mergedConfig.env;
	return mergedConfig;
}

/**
 * Detects basic config options to use if the .pbrocks-env.json config file does not
 * exist. For example, if the local directory contains a plugin, that will be
 * added to the default plugin sources.
 *
 * @param {string} configPath A path to the config file for the source to detect.
 * @return {Object} Basic config options for the detected source type. Empty
 *                  object if no config detected.
 */
async function getDefaultBaseConfig( configPath ) {
	const configDirectoryPath = path.dirname( configPath );
	const type = await detectDirectoryType( configDirectoryPath );

	if ( type === 'core' ) {
		return { core: '.' };
	} else if ( type === 'plugin' ) {
		return { plugins: [ '.' ] };
	} else if ( type === 'theme' ) {
		return { themes: [ '.' ] };
	}

	return {};
}

/**
 * Overrides keys in the config object with set environment variables or options
 * which should be merged.
 *
 * @param {WPConfig} config fully parsed configuration object.
 * @return {WPConfig} configuration object with overrides applied.
 */
function withOverrides( config ) {
	const workDirectoryPath = config.workDirectoryPath;
	// Override port numbers with environment variables.
	config.env.development.port =
		getNumberFromEnvVariable( 'WP_ENV_PORT' ) ||
		config.env.development.port;
	config.env.tests.port =
		getNumberFromEnvVariable( 'WP_ENV_TESTS_PORT' ) ||
		config.env.tests.port;

	// Override WordPress core with environment variable.
	if ( process.env.WP_ENV_CORE ) {
		const coreSource = includeTestsPath(
			parseSourceString( process.env.WP_ENV_CORE, { workDirectoryPath } ),
			{ workDirectoryPath }
		);
		config.env.development.coreSource = coreSource;
		config.env.tests.coreSource = coreSource;
	}

	// Override PHP version with environment variable.
	config.env.development.phpVersion =
		process.env.WP_ENV_PHP_VERSION || config.env.development.phpVersion;
	config.env.tests.phpVersion =
		process.env.WP_ENV_PHP_VERSION || config.env.tests.phpVersion;

	const updateEnvUrl = ( configKey ) => {
		[ 'development', 'tests' ].forEach( ( envKey ) => {
			try {
				const baseUrl = new URL(
					config.env[ envKey ].config[ configKey ]
				);

				// Don't overwrite the port of WP_HOME when set.
				if ( ! ( configKey === 'WP_HOME' && !! baseUrl.port ) ) {
					baseUrl.port = config.env[ envKey ].port;
				}

				config.env[ envKey ].config[ configKey ] = baseUrl.toString();
			} catch ( error ) {
				throw new ValidationError(
					`Invalid .pbrocks-env.json: config.${ configKey } must be a valid URL.`
				);
			}
		} );
	};

	// Update wp config options to include the correct port number in the URL.
	updateEnvUrl( 'WP_SITEURL' );
	updateEnvUrl( 'WP_HOME' );

	return config;
}

/**
 * Parses an environment variable which should be a number.
 *
 * Throws an error if the variable cannot be parsed to a number.
 * Returns null if the environment variable has not been specified.
 *
 * @param {string} varName The environment variable to check (e.g. WP_ENV_PORT).
 * @return {null|number} The number. Null if it does not exist.
 */
function getNumberFromEnvVariable( varName ) {
	// Allow use of the default if it does not exist.
	if ( ! process.env[ varName ] ) {
		return null;
	}

	const maybeNumber = parseInt( process.env[ varName ] );

	// Throw an error if it is not parseable as a number.
	if ( isNaN( maybeNumber ) ) {
		throw new ValidationError(
			`Invalid environment variable: ${ varName } must be a number.`
		);
	}

	return maybeNumber;
}

/**
 * Gets the `pbrocks-env` home directory in which generated files are created.
 *
 * By default: '~/.pbrocks-env/'. On Linux with snap packages: '~/pbrocks-env/'. Can be
 * overriden with the WP_ENV_HOME environment variable.
 *
 * @return {Promise<string>} The absolute path to the `pbrocks-env` home directory.
 */
async function getHomeDirectory() {
	// Allow user to override download location.
	if ( process.env.WP_ENV_HOME ) {
		return path.resolve( process.env.WP_ENV_HOME );
	}

	/**
	 * Installing docker with Snap Packages on Linux is common, but does not
	 * support hidden directories. Therefore we use a public directory when
	 * snap packages exist.
	 *
	 * @see https://github.com/WordPress/gutenberg/issues/20180#issuecomment-587046325
	 */
	return path.resolve(
		os.homedir(),
		!! ( await fs.stat( '/snap' ).catch( () => false ) )
			? 'pbrocks-env'
			: '.pbrocks-env'
	);
}
